common/workunit/workunit.cpp

```diff
@@ -3035,6 +3041,117 @@ ScopeCompare WuScopeFilter::compareMatchScopes(const char * scope) const
 }
 
 
+StringBuffer & WuScopeFilter::describe(StringBuffer & out) const
+{
+    scopeFilter.describe(out);
+    if (requiredStats.size())
+    {
+        out.append(",where[");
+        bool first = false;
+        for (const auto & stat : requiredStats)
+        {
+            if (!first)
+                out.append(",");
+            stat.describe(out);
+            first = false;
+        }
+        out.append("]");
+    }
+
+    {
+        StringBuffer sources;
+        for (unsigned mask=1; mask; mask *= 2)
+        {
+            if (sourceFlags & mask)
+            {
+                const char * source = querySourceText((WuScopeSourceFlags)mask);
+                if (source)
+                    sources.append(",").append(source);
+            }
+        }
+        if (sources)
+            out.append(",source[").append(sources.str()+1).append("]");
+    }
+
+    if (include.nestedDepth != UINT_MAX)
+        out.appendf(",nested[%u]", include.nestedDepth);
+
+    if (include.scopeTypes)
+    {
+        out.append(",include[");
+        ForEachItemIn(i, include.scopeTypes)
+        {
+            if (i)
+                out.append(",");
+            out.append(queryScopeTypeName((StatisticScopeType)include.scopeTypes.item(i)));
+        }
+        out.append("]");
+    }
+
+    {
+        StringBuffer props;
+        if (properties == PTnone)
+            props.append(",none");
+        else if (properties == PTall)
+            props.append(",all");
+        else
+        {
+            if (properties & PTstatistics)
+                props.append(",stat");
+            if (properties & PTattributes)
+                props.append(",attr");
+            if (properties & PThints)
+                props.append(",hint");
+            if (properties & PTscope)
+                props.append(",scope");
+        }
+        out.append(",properties[").append(props.str()+1).append("]");
+    }
+
+    if (desiredStats)
+    {
+        out.append(",stat[");
+        ForEachItemIn(i, desiredStats)
+        {
+            if (i)
+                out.append(",");
+            out.append(queryStatisticName((StatisticKind)desiredStats.item(i)));
+        }
+        out.append("]");
+    }
+
+    if (desiredAttrs)
+    {
+        out.append(",attr[");
+        ForEachItemIn(i, desiredAttrs)
+        {
+            if (i)
+                out.append(",");
+            out.append(queryWuAttributeName((WuAttr)desiredAttrs.item(i)));
+        }
+        out.append("]");
+    }
+
+    if (desiredHints)
+    {
+        out.append(",hint[");
+        ForEachItemIn(i, desiredHints)
+        {
+            if (i)
+                out.append(",");
+            out.append(desiredHints.item(i));
+        }
+        out.append("]");
+    }
+
+    if (desiredMeasure != SMeasureAll)
+        out.append(",measure[").append(queryMeasureName(desiredMeasure)).append("]");
+
+    if (minVersion != 0)
+        out.appendf(",version(%" I64F "u)", minVersion);
+
+    return out;
```

Comment:
Should a string with a leading comma be returned here?

Reply:
This is only used for tracing, and the caller could easily skip the leading ',' if it was not wanted.  I opted for simplicity rather than trying to suppress the leading ','